lib/workflow/task-graph.js

```diff
@@ -0,0 +1,493 @@
+        assert.string(baseTaskDefinition.runJob, 'runJob');
+        assert.object(baseTaskDefinition.requiredOptions, 'requiredOptions');
+        assert.object(baseTaskDefinition.requiredProperties, 'requiredProperties');
+        assert.object(baseTaskDefinition.properties, 'properties');
+    };
+
+    TaskGraph.prototype._validateProperties = function(taskDefinition, context) {
+        var self = this;
+        var baseTaskDefinition = self._getBaseTask(taskDefinition);
+        var requiredProperties = baseTaskDefinition.requiredProperties;
+        _.forEach(requiredProperties, function(v, k) {
+            self.compareNestedProperties(
+                v, k, context.properties, baseTaskDefinition.injectableName);
+        });
+
+        // Update shared context with properties from this task
+        var _properties = _.merge(taskDefinition.properties, baseTaskDefinition.properties);
+        context.properties = _.merge(_properties, context.properties);
+    };
+
+    TaskGraph.prototype._validateOptions = function(taskDefinition, label) {
+        var self = this;
+        var baseTaskDefinition = self._getBaseTask(taskDefinition);
+        _.forEach(baseTaskDefinition.requiredOptions, function(k) {
+            var option = taskDefinition.options[k];
+            if (!option && _.has(self.definition.options.defaults, k)) {
+                option = self.definition.options.defaults[k];
+            }
+            if (label && _.has(self.definition.options[label], k)) {
+                option = self.definition.options[label][k];
+            }
+            assert.ok(option,
+                'required option ' + k + ' for task ' +
+                taskDefinition.injectableName + ' in graph ' + self.injectableName);
+        });
+    };
+
+    TaskGraph.prototype._getBaseTask = function(definition) {
+        assert.object(definition);
+        assert.string(definition.implementsTask);
+
+        // TODO: this is just a temporary solution until base tasks are refactored
+        // to be attributes of the job classes themselves.
+        // TODO: also un-promisify all _getBaseTask calls *again* :(
+        var baseTaskDefinition = _.find(taskLibrary, function(task) {
+            return !task.implementsTask && task.injectableName === definition.implementsTask;
+        });
+        assert.object(baseTaskDefinition, "Base task definition for " +
+                definition.implementsTask + " should exist");
+        return baseTaskDefinition;
+    };
+
+    TaskGraph.prototype.compareNestedProperties = function(value, nestedKey, obj, taskName) {
+        var self = this;
+
+        // nested key is a dot notated string that represents a JSON scope, e.g.
+        // os.linux.type represents { os: { linux: { type: 'value' } } }
+        if (!nestedKey) {
+            return;
+        }
+        assert.string(nestedKey);
+        var keys = nestedKey.split('.');
+        if (keys.length === 1) {
```

Comment:
Could lodash .get be used here?  Assert (value === _.get (obj, nestedKey, null))